src/main/java/org/prebid/server/bidder/pulsepoint/PulsepointBidder.java

```diff
@@ -1,89 +1,178 @@
+        String publisherId = null;
+
+        for (Imp imp : bidRequest.getImp()) {
+            final ExtImpPulsepoint extImpPulsepoint;
+            try {
+                extImpPulsepoint = parseImpExt(imp);
+            } catch (PreBidException e) {
+                errors.add(BidderError.badInput(e.getMessage()));
+                continue;
+            }
+
+            final Imp modifiedImp = modifyImp(imp, extImpPulsepoint);
+            modifiedImps.add(modifiedImp);
+
+            final Integer extPublisherId = extImpPulsepoint.getPublisherId();
+            if (publisherId == null && extPublisherId != null && extPublisherId > 0) {
+                publisherId = extPublisherId.toString();
+            }
+        }
+
+        if (modifiedImps.isEmpty()) {
+            return Result.withErrors(errors);
+        }
+        publisherId = Objects.requireNonNullElse(publisherId, StringUtils.EMPTY);
+
+        final BidRequest modifiedRequest = modifyRequest(bidRequest, publisherId, modifiedImps);
+        return Result.of(Collections.singletonList(createHttpRequest(modifiedRequest)), errors);
+    }
+
+    private ExtImpPulsepoint parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), PULSEPOINT_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage());
+        }
+    }
+
+    private static Imp modifyImp(Imp imp, ExtImpPulsepoint extImpPulsepoint) {
         return imp.toBuilder().tagid(Objects.toString(extImpPulsepoint.getTagId())).build();
     }
 
-    @Override
-    protected void modifyRequest(BidRequest bidRequest, BidRequest.BidRequestBuilder requestBuilder,
-                                 List<ImpWithExt<ExtImpPulsepoint>> impsWithExts) {
-        final String pubId = impsWithExts.stream()
-                .map(ImpWithExt::getImpExt)
-                .map(ExtImpPulsepoint::getPublisherId)
-                .filter(this::isValidPublisherId)
-                .findFirst()
-                .map(Objects::toString)
-                .orElse(StringUtils.EMPTY);
-
-        final Site site = bidRequest.getSite();
-        final App app = bidRequest.getApp();
+    private static BidRequest modifyRequest(BidRequest request, String publisherId, List<Imp> imps) {
+        final BidRequest.BidRequestBuilder builder = request.toBuilder();
+
+        final Site site = request.getSite();
+        final App app = request.getApp();
         if (site != null) {
-            requestBuilder.site(modifySite(site, pubId));
+            builder.site(modifySite(site, publisherId));
         } else if (app != null) {
-            requestBuilder.app(modifyApp(app, pubId));
+            builder.app(modifyApp(app, publisherId));
         }
-    }
+        builder.imp(imps);
 
-    private boolean isValidPublisherId(Integer publisherId) {
-        return publisherId != null && publisherId > 0;
+        return builder.build();
```

Comment:
I believe that this logic can be rewritten to be more declarative, if you slightly modify `modifySite` and `modifyApp` methods.
It will look smth like:
```
request.toBuilder()
   .app(modifyApp(..))
   .site(modifySite(..))
   .imp(imps)
   .build();
```
Also, keep in mind that Site and App can't be present at the same time, as core logic will remove Site if both present.
See: https://github.com/prebid/prebid-server-java/blob/c6a298f674bcc44164d6337798499e297336658b/src/main/java/org/prebid/server/validation/RequestValidator.java#L178